litr/src/main/java/com/linkedin/android/litr/io/AudioRecordMediaSource.kt

```diff
@@ -0,0 +1,179 @@
+package com.linkedin.android.litr.io
+
+import android.media.*
+import android.os.Build
+import android.util.Log
+import androidx.annotation.RequiresApi
+import com.linkedin.android.litr.MimeType
+import com.linkedin.android.litr.exception.MediaSourceException
+import java.nio.ByteBuffer
+
+/**
+ * An implementation of MediaSource, which utilizes Android's {@link AudioRecord} to record audio
+ * from a given audio source (i.e. the device's microphone).
+ *
+ * This class requires Android M+ in order to support reading data from {@link AudioRecord} in a
+ * non-blocking way.
+ */
+@RequiresApi(Build.VERSION_CODES.M)
+class AudioRecordMediaSource(
+        private val audioSource: Int = DEFAULT_AUDIO_SOURCE,
+        private val sampleRate: Int = DEFAULT_SAMPLE_RATE,
+        private val channelConfig: Int = DEFAULT_CHANNEL_CONFIG,
```

Comment:
It feels like it is possible to easily mismatch `channelConfig` and `channelCount`, which would cause problems, I presume. Should we just accept `channelConfig` and derive `channelCount` from it?

Reply:
Something like this:

```
class AudioRecordMediaSource(
        private val audioSource: Int = DEFAULT_AUDIO_SOURCE,
        private val sampleRate: Int = DEFAULT_SAMPLE_RATE,
        private val isStereo: Boolean = false
) : MediaSource {

    private val channelCount: Int by lazy { 
        return@lazy if (isStereo) 2 else 1
    }
    
    private val channelConfig: Int by lazy {
        return@lazy if (isStereo) AudioFormat.CHANNEL_IN_STEREO else AudioFormat.CHANNEL_IN_MONO
    }
```

Reply:
In that case I would probably keep the parameter an integer `channelCount` - masks internal `AudioConfig` implementation and is more future-proof (who knows, maybe someday Android devices will record a 6 channel sound?)

Reply:
Sure, will tweak 👍